### PR TITLE
Add python package

### LINKED ISSUE #166: Simplify installation
Would it make sense to make the collection more easily installable, e.g. as a python package via pypi?

Advantages:
* Ease of installation.
* Easy to maintain only a single collection on a system, if installation path is globally registered.
* Could provide access functions (overview, list of models, ...)

### PR DESCRIPTION
**Pull request by yannikschaelte**
* Add python package for easier installation, fixes #166 

OLD, UPDATED BELOW:
- [x] Someone with admin rights would need to upload secrets for twine for auto-deploy
- Upon model additions/edits, new releases would need to be created on pypi to ensure consistency. Possible to automate?~~

### CODE REVIEW

File: src/python/setup.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+def read(fname):
+    with open(fname) as f:
+        return f.read()
+
+name = "benchmark_models_petab"
+
+setuptools.setup(
+    # Metadata
+    name=name,
+    version="0.0.0a2",
```
Comment (dweindl) [inline src/python/setup.py:18]: Store somewhere else and set `benchmark_models_petab.__version__` accordingly?

File: src/python/build.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+# Remove link
+rm -rf $BUILD_DIR/$CODE_DIR/$DATA_DIR
+# Copy data
+cp -r "../../$DATA_DIR" $BUILD_DIR/$CODE_DIR
+
+# Hop into
+cd $BUILD_DIR
+
+# Build
+python setup.py sdist
```
Comment (dweindl) [inline src/python/build.sh:30]: Using setup.py directly is deprecated. Use `python3 -m build --sdist` instead.

File: .github/workflows/python.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    - name: Prepare python ${{ matrix.python-version }}
+      uses: actions/setup-python@v1
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Run tests
+      run: |
+        cd src/python
+        pip install --upgrade pip
+        pip install -e .
```
Comment (dweindl) [inline .github/workflows/python.yml:27]: Ideally install from an sdist and try to load a PEtab problem, to ensure all data files have been added to the package correctly.

File: src/python/benchmark_models_petab/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+from .C import (
+    MODELS_DIR,
+    MODEL_DIRS,
+    MODELS,
+)
```
Comment (dweindl) [inline src/python/benchmark_models_petab/__init__.py:5]: I could imagine some `get_problem(id_: str) -> petab.Problem` would be handy...

File: src/python/build.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+# Create build folder
+if [ -e $BUILD_DIR ]; then
+    echo "Remove build folder $BUILD_DIR"
+    exit 1
+fi
+mkdir $BUILD_DIR
+
+# Copy code
+cp -r \
+  $CODE_DIR "setup.py" "MANIFEST.in" "../../LICENSE" "../../README.md" \
```
Comment (dweindl) [inline src/python/build.sh:19]: Assumes it's running from `src/python/`. Please change all paths to absolute paths, based on the script location.

**Comment by dweindl** [comment]
> * [ ]  Someone with admin rights would need to upload secrets for twine for auto-deploy
> 
> * Upon model additions/edits, new releases would need to be created on pypi to ensure consistency. Possible to automate?

Do we really need it on pypi? Could just pip install it from github? Saves us from changing the version number in every commit.

**Comment by yannikschaelte** [comment]
> > * [ ]   Someone with admin rights would need to upload secrets for twine for auto-deploy
> > * Upon model additions/edits, new releases would need to be created on pypi to ensure consistency. Possible to automate?
> 
> Do we really need it on pypi? Could just pip install it from github? Saves us from changing the version number in every commit.

@dweindl I removed the pypi deploy. However, while `pip install git+...` is possible, I'm not sure how an install from git with source would work?

**Comment by dweindl** [comment]
> I'm not sure how an install from git with source would work?

Ah, right, won't work with that directory structure :(. Would have to move everything into the package directory. Wouldn't do that. 

:thinking:

**Comment by dweindl** [comment]
> > I'm not sure how an install from git with source would work?
> 
> Ah, right, won't work with that directory structure :(. Would have to move everything into the package directory. Wouldn't do that.
> 
> thinking

Ah no, at least on Linux `pip install git+https://github.com/Benchmarking-Initiative/Benchmark-Models-PEtab.git@fixes_ys#subdirectory=src/python` works.


Could somebody tests this on Windows?
```shell
pip install git+https://github.com/Benchmarking-Initiative/Benchmark-Models-PEtab.git@fixes_ys#subdirectory=src/python
python -c "from benchmark_models_petab import get_problem; print(get_problem('Zheng_PNAS2012').get_x_ids())"
```
I guess it only works if symlinks are enabled?!

**Comment by yannikschaelte** [comment]
> > > I'm not sure how an install from git with source would work?
> > 
> > 
> > Ah, right, won't work with that directory structure :(. Would have to move everything into the package directory. Wouldn't do that.
> > thinking
> 
> Ah no, at least on Linux `pip install git+https://github.com/Benchmarking-Initiative/Benchmark-Models-PEtab.git@fixes_ys#subdirectory=src/python` works.
> 
> Could somebody tests this on Windows?
> 
> ```shell
> pip install git+https://github.com/Benchmarking-Initiative/Benchmark-Models-PEtab.git@fixes_ys#subdirectory=src/python
> python -c "from benchmark_models_petab import get_problem; print(get_problem('Zheng_PNAS2012').get_x_ids())"
> ```
> 
> I guess it only works if symlinks are enabled?!

Windows testing would take some effort. Maybe just assume we'll be notified when this becomes a problem?

**Review by dweindl — APPROVED** [review]
:+1:

File: src/python/setup.cfg [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+[metadata]
+name = benchmark_models_petab
+version = attr: benchmark_models_petab.version.__version__
+description = "A collection of models with experimental data in the PEtab format"
+long_description = file: README.md
+long_description_content_type = text/markdown
+url = "https://github.com/Benchmarking-Initiative/Benchmark-Models-PEtab"
+
+author = "The Benchmark-Models-PEtab maintainers"
+author_email = "yannik.schaelte@gmail.com"
```
Comment (yannikschaelte) [inline src/python/setup.cfg:12]: happy to replace by any other mail ;)

**Pull request merged**